File: compress_tool.py
from pathlib import Path
import zipfile
from check_file_permission import check_file_permission

def compress_files(file_paths, zip_path, config):
    if not check_file_permission(zip_path, "write", config):
        raise PermissionError(f"Write access denied for {zip_path}")
    zip_path = Path(zip_path).resolve()
    with zipfile.ZipFile(zip_path, "w", zipfile.ZIP_DEFLATED) as zf:
        for file_path in file_paths:
            if not check_file_permission(file_path, "read", config):
                raise PermissionError(f"Read access denied for {file_path}")
            file_path = Path(file_path).resolve()
            zf.write(file_path, file_path.name)
    return f"Compressed to {zip_path}"

File: ts-tools/compress_tool.ts
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import * as fs from 'fs';
import * as path from 'path';
import * as archiver from 'archiver';
import { checkFilePermission } from './check_file_permission';

export function compressFiles(
  filePaths: string[],
  zipPath: string,
  config: any,
): Promise<string> {
  return new Promise((resolve, reject) => {
    checkFilePermission(zipPath, 'write', config)
      .then(() => {
        const resolvedZipPath = path.resolve(zipPath);
        const output = fs.createWriteStream(resolvedZipPath);
        const archive = archiver('zip', {
          zlib: { level: 9 },
        });

        output.on('close', () => {
          resolve(`Compressed to ${zipPath}`);
        });

        archive.on('error', (err) => {
          reject(err);
        });

        archive.pipe(output);

        Promise.all(
          filePaths.map((filePath) =>
            checkFilePermission(filePath, 'read', config).then(() => {
              const resolvedPath = path.resolve(filePath);
              archive.file(resolvedPath, { name: path.basename(resolvedPath) });
            }),
          ),
        )
          .then(() => archive.finalize())
          .catch(reject);
      })
      .catch(reject);
  });
}
